biobalm/_sd_attractors/attractor_candidates.py

```diff
@@ -0,0 +1,760 @@
+
+        # This tests the assumption that all fully expanded minimal
+        # traps are handled before, and the only trap spaces with
+        # an empty NFVS are unexpanded pseudo-minimal,
+        # or expanded non-minimal.
+        assert not sd.node_is_minimal(node_id)
+
+        if not node_is_pseudo_minimal:
+            if biobalm.succession_diagram.DEBUG:
+                print(
+                    f"[{node_id}] > Attractor candidates done: empty NFVS in a non-minimal space."
+                )
+            return []
+
+    # First, we create a retained set and corresponding candidate set that is valid in the
+    # percolated Boolean network. This process is *usually* quite fast, but
+    # sometimes the candidate set explodes and is very large (even too large to enumerate).
+    # In that case, we can try to simplify it greedily using the ASP solver (assuming this
+    # is enabled).
+    #
+    # Overall, our strategy is the following:
+    # 1. Compute a heuristic retained set based on the original NFVS method.
+    # 2. If the ASP optimization is disabled, we generate the candidates from this retained
+    #    set in full.
+    # 3. If ASP optimization is enabled, we first test if there are at least 100 candidates
+    #    for the initial retained set.
+    #    3a. If there are fewer than 100 candidates, we just run one optimization pass on
+    #        the initial retained set and report the result.
+    #    3b. If there are more than 100 candidates, we probably need to optimize the retained
+    #        set even further, in which case we start to recursively generate a new retained
+    #        set that is optimized in each step to hopefully get a better result.
+
+    pn_reduced = sd.node_percolated_petri_net(node_id, compute=True)
+    bn_reduced = sd.node_percolated_network(node_id, compute=True)
+    graph_reduced = AsynchronousGraph(bn_reduced)
+
+    retained_set = make_heuristic_retained_set(
+        graph_reduced, node_nfvs, child_motifs_reduced
+    )
+
+    if len(retained_set) == sd.network.variable_count() and node_is_pseudo_minimal:
+        # If the retained set describes a fixed point, then only one attractor
+        # is present in this space and it must contain the state described by the retained set.
+        if biobalm.succession_diagram.DEBUG:
+            print(
+                f"[{node_id}] > Singular attractor found through fixed-point retained set. Done."
+            )
+        return [retained_set | node_space]
+
+    if not greedy_asp_minification:
+        candidate_states = compute_fixed_point_reduced_STG(
+            pn_reduced,
+            retained_set,
+            avoid_subspaces=child_motifs_reduced,
+            solution_limit=CANDIDATES_HARD_LIMIT,
+        )
+        if len(candidate_states) == CANDIDATES_HARD_LIMIT:
+            raise RuntimeError(
+                f"Exceeded the maximum amount of attractor candidates (CANDIDATES_HARD_LIMIT={CANDIDATES_HARD_LIMIT})."
```

Comment:
not a big deal, but you can write this as `f"Exceeded the maximum amount of attractor candidates ({CANDIDATES_HARD_LIMIT=})."`